File: 2109-adding-spaces-to-a-string/2109-adding-spaces-to-a-string.c
#include <stdlib.h>
#include <string.h>

/**
 * @brief Adds spaces to the string `s` at the specified indices in `spaces`.
 *
 * @param s (char*): The original string.
 * @param spaces (int*): The list of indices where spaces should be added.
 * @param spacesSize (int): The size of the `spaces` array.
 * @return (char*): The modified string with spaces added.
 */
char* addSpaces(char* s, int* spaces, int spacesSize) {
    int sLen = strlen(s);
    int newLen = sLen + spacesSize;  // New length of the string with spaces
    char* result = (char*)malloc((newLen + 1) * sizeof(char));  // Allocate memory for the new string

    int sIdx = 0, spaceIdx = 0, resIdx = 0;

    // Traverse the original string `s` and add spaces where needed
    while (sIdx < sLen) {
        // Check if the current index matches a space position
        if (spaceIdx < spacesSize && sIdx == spaces[spaceIdx]) {
            result[resIdx++] = ' ';  // Add a space
            spaceIdx++;  // Move to the next space index
        }
        result[resIdx++] = s[sIdx++];  // Add the current character
    }

    result[resIdx] = '\0';  // Null-terminate the result string
    return result;  // Return the modified string
}

File: 2109-adding-spaces-to-a-string/2109-adding-spaces-to-a-string.py
from typing import List

class Solution:
    def addSpaces(self, s: str, spaces: List[int]) -> str:
        """
        Adds spaces to the string `s` at the specified indices in `spaces`.

        Parameters:
        s (str): The original string.
        spaces (List[int]): The list of indices where spaces should be added.

        Returns:
        str: The modified string with spaces added.
        """
        result = []  # List to build the result efficiently
        space_idx = 0  # Pointer for the spaces array
        n = len(spaces)
        
        # Traverse the string `s`
        for i, char in enumerate(s):
            # If the current index matches the space position
            if space_idx < n and i == spaces[space_idx]:
                result.append(' ')  # Add a space
                space_idx += 1  # Move to the next space position
            
            result.append(char)  # Add the current character
        
        return ''.join(result)  # Convert list to string
